src/Pact/Analyze/Eval/Term.hs

```diff
@@ -611,54 +611,75 @@ evalTerm = \case
       Some (SObject _) _ -> throwErrorNoLoc "We can't yet analyze calls to `hash` on objects"
       Some _ _           -> throwErrorNoLoc "We can't yet analyze calls to `hash` on non-{string,integer,bool}"
 
-  Step mEntity (tm :< ty) mRollback mContinue -> do
-    case mEntity of
-      Nothing -> pure ()
-      Just expectedEntity -> do
-        actualEntity    <- view currentEntity
-        expectedEntity' <- eval expectedEntity
-        markFailure $ actualEntity ./= expectedEntity'
-
-    case mRollback of
-      Nothing               -> pure ()
-      Just (rollback, _tid) -> globalState . gasRollbacks %= (rollback:)
-
-    _tmVal <- withSing ty $ eval tm
-    case mContinue of
-      Nothing       -> pure "step done"
-      Just continue -> eval continue
-
-
-  IntraStepReset failureVid tm -> do
-    rollbacks <- use $ globalState . gasRollbacks
-    tables    <- view $ analyzeEnv . aeTables
-    oldState  <- use id
-    oldEnv    <- ask
-    let txMetadata   = TxMetadata (mkFreeArray "txKeySets")
-                                  (mkFreeArray "txDecimals")
-                                  (mkFreeArray "txIntegers")
-
-        newState = oldState
-          & latticeState . lasExtra . cvTableCells .~ mkSymbolicCells tables
-          -- & latticeState . lasExtra . cvRowExists .~  i think this one is okay
-        newEnv = oldEnv
-          & aePactMetadata . pmEntity .~ uninterpretS "entity"
-          -- & registry                  .~ undefined
-          & aeTxMetadata              .~ txMetadata
-          -- & aeScope                   .~ undefined
-          -- & aeGuardPasses             .~ undefined
-
-    view (aeNondets . at failureVid) >>= \case
-      Nothing      -> throwError undefined
-      Just failure -> do
-        id .= newState
-
-        -- both continue and roll back via nondeterminism
-        local (analyzeEnv .~ newEnv) $ ite failure
-          -- (return value never used)
-          (evalTerm tm             >> pure "step done")
-          (for rollbacks evalETerm >> pure "step done")
-
+  Pact steps -> do
+    _ <- foldlM
+      (\(rollbacks, afterFirstStep)
+        (Step mEntity (tm :< ty) mRollback) -> withSing ty $ do
+
+            -- We reset the environment unless this is the first step
+        let withReset' = if afterFirstStep then withReset else id
+
+        _ <- withReset' $ case mRollback of
+
+          -- If there's no rollback then we just evaluate this term
+          Nothing -> void $ evalTermWithEntity mEntity tm
+
+          -- ... otherwise, we nondeterministically either execute the step or
+          -- all the existing rollbacks. Note that we use 'rollbacks', ie all
+          -- rollbacks prior to this step, not 'rollbacks'', because we don't
+          -- want to execute the rollback associated with this step.
+          Just (_rollback, failureVid) -> do
+            failure <- view (aeNondets . at failureVid)
+              ??? "couldn't find failure var"
+
+            ite failure
```

Comment:
maybe we should adjust this name? it's a little confusing that we proceed with normal step evaluation in the case of "failure"